catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/DumpCommand.java

```diff
@@ -240,69 +245,32 @@ protected Object executeWithSubject() throws Exception {
             rejectedExecutionHandler);
 
-    while (!response.getResults().isEmpty()) {
-      response =
-          new SourceResponseImpl(
-              new QueryRequestImpl(query, props),
-              ResultIterable.resultIterable(
-                      catalog::query, new QueryRequestImpl(query, props), pageSize)
-                  .stream()
-                  .collect(Collectors.toList()));
-
-      if (StringUtils.isNotBlank(zipFileName)) {
-        try {
-          Optional<QueryResponseTransformer> zipCompressionTransformer = getZipCompression();
-          if (zipCompressionTransformer.isPresent()) {
-            BinaryContent binaryContent =
-                zipCompressionTransformer.get().transform(response, zipArgs);
-            if (binaryContent != null) {
-              IOUtils.closeQuietly(binaryContent.getInputStream());
-            }
-            Long resultSize = (long) response.getResults().size();
-            printStatus(resultCount.addAndGet(resultSize));
-          }
-        } catch (InvalidSyntaxException e) {
-          LOGGER.info("No Zip Transformer found.  Unable export metacards to a zip file.");
-        } catch (CatalogTransformerException e) {
-          LOGGER.info("zipCompression transform failed");
-        }
-      } else if (multithreaded > 1) {
-        final List<Result> results = new ArrayList<>(response.getResults());
-        executorService.submit(
-            () -> {
-              for (final Result result : results) {
-                Metacard metacard = result.getMetacard();
-                try {
-                  exportMetacard(dumpDir, metacard, resultCount);
-                } catch (IOException e) {
-                  LOGGER.debug("Failed to dump metacard {}", metacard.getId(), e);
-                }
-              }
-            });
-      } else {
-        for (final Result result : response.getResults()) {
-          Metacard metacard = result.getMetacard();
-          exportMetacard(dumpDir, metacard, resultCount);
-        }
-      }
-
-      if (response.getResults().size() < pageSize || pageSize == -1) {
-        break;
-      }
-
-      query.setStartIndex(query.getStartIndex() + pageSize);
+    QueryRequest queryRequest = new QueryRequestImpl(query, props);
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug("Hits for Search: {}", catalog.query(queryRequest).getHits());
     }
 
+    ResultIterable.resultIterable(catalog::query, queryRequest)
+        .stream()
+        .map(Collections::singletonList)
+        .map(result -> new SourceResponseImpl(queryRequest, result))
+        .forEach(response -> handleResult(response, executorService, dumpDir, resultCount));
+
     executorService.shutdown();
 
+    boolean interrupted = false;
     while (!executorService.isTerminated()) {
       try {
         TimeUnit.MILLISECONDS.sleep(100);
       } catch (InterruptedException e) {
-        // ignore
+        interrupted = true;
       }
     }
 
+    if (interrupted) {
```

Comment:
❗️ This must be in a `try/finaly` where the `try` surrounds the while loop and this if is within the `finally`

Reply:
Why is it necessary to double nest the while loop in 2 try blocks? Shouldn't the try/catch handle the exception being thrown? This code will still execute after the executor service completes.

Reply:
Because if there are any exceptions generated from the next loop after an interruption was detected, the interruption state doesn't get propagated.

Typically you would do:

```
boolean interrupted = false;

try {
  while (!executorService.isTerminated()) {
    try {
      TimeUnit.MILLISECONDS.sleep(100);
    } catch (InterruptedException e) {
      interrupted = true;
    }
  }
} finally {
  if (interrupted) {
    Thread.currentThread().interrupt();
  }
}
```

Reply:
Ok. Wrapped the while loop in try/finally in case executorService.isTerminated or sleep throws a RuntimeException